File: Javascript/Best_Time_To_Buy_And_Sell_Stock.js
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
// Solution By Emmanuel Ezenwigbo
/* Time Complexity - O(n)
   Space Complexity - O(1)
*/

var maxProfit = function(prices) {
    let [min_price, maxP] = [Infinity,0];
    
    for (let i = 0; i < prices.length; i++){
        max_sell = prices[i] - min_price
        maxP = Math.max(maxP, max_sell)
        min_price = Math.min(min_price, prices[i])
    }
    
    return maxP;
};

File: C++/Best_time_to_buy_and_sell_stock.cpp
class Solution {
public:
    long  maxprofit(int ind,int buy,vector<int>& prices,vector<vector<int>>& dp)
    {
        if(ind==prices.size())
        {
            return 0;
        }
        if(dp[ind][buy]!=-1)
        {
            return dp[ind][buy];
        }
        long profit=0;
        if(buy)
        {
            profit=max(-prices[ind]+maxprofit(ind+1,0,prices,dp),maxprofit(ind+1,1,prices,dp));
        }
        else
        {
            profit=max(prices[ind]+maxprofit(ind+1,1,prices,dp),maxprofit(ind+1,0,prices,dp));
        }
        return dp[ind][buy]=profit;
    }
    int maxProfit(vector<int>& prices) {
        int n=prices.size();
        vector<vector<int>>dp(n,vector<int>(2,-1));
        return maxprofit(0,1,prices,dp);
    }
    
    
};
